buffer/src/main/java/io/netty/buffer/SizeClasses.java

```diff
@@ -0,0 +1,403 @@
+
+    private int[] pageIdx2sizeTab;
+
+    // lookup table for sizeIdx <= smallMaxSizeIdx
+    private int[] sizeIdx2sizeTab;
+
+    // lookup table used for size <= lookupMaxclass
+    // spacing is 1 << LOG2_QUANTUM, so the size of array is lookupMaxclass >> LOG2_QUANTUM
+    private int[] size2idxTab;
+
+    private void sizeClasses() {
+        int normalMaxSize = -1;
+
+        int index = 0;
+        int size = 0;
+
+        int log2Group = LOG2_QUANTUM;
+        int log2Delta = LOG2_QUANTUM;
+        int ndeltaLimit = 1 << LOG2_SIZE_CLASS_GROUP;
+
+        //First small group, nDelta start at 0.
+        //first size class is 1 << LOG2_QUANTUM
+        int nDelta = 0;
+        while (nDelta < ndeltaLimit) {
+            size = sizeClass(index++, log2Group, log2Delta, nDelta++);
+        }
+        log2Group += LOG2_SIZE_CLASS_GROUP;
+
+        //All remaining groups, nDelta start at 1.
+        while (size < chunkSize) {
+            nDelta = 1;
+
+            while (nDelta <= ndeltaLimit && size < chunkSize) {
+                size = sizeClass(index++, log2Group, log2Delta, nDelta++);
+                normalMaxSize = size;
+            }
+
+            log2Group++;
+            log2Delta++;
+        }
+
+        //chunkSize must be normalMaxSize
+        assert chunkSize == normalMaxSize;
+
+        nSizes = index;
+    }
+
+    //calculate size class
+    private int sizeClass(int index, int log2Group, int log2Delta, int nDelta) {
+        short isMultiPageSize;
+        if (log2Delta >= pageShifts) {
+            isMultiPageSize = yes;
+        } else {
+            int pageSize = 1 << pageShifts;
+            int size = (1 << log2Group) + (1 << log2Delta) * nDelta;
+
+            isMultiPageSize = size == size / pageSize * pageSize? yes : no;
+        }
+
+        int log2Ndelta = nDelta == 0? 0 : log2(nDelta);
+
+        byte remove = 1 << log2Ndelta < nDelta? yes : no;
+
+        int log2Size = log2Delta + log2Ndelta == log2Group? log2Group + 1 : log2Group;
+        if (log2Size == log2Group) {
+            remove = yes;
+        }
+
+        short isSubpage = log2Size < pageShifts + LOG2_SIZE_CLASS_GROUP? yes : no;
+
+        int log2DeltaLookup = log2Size < LOG2_MAX_LOOKUP_SIZE ||
+                              log2Size == LOG2_MAX_LOOKUP_SIZE && remove == no
+                ? log2Delta : no;
+
+        short[] sz = {
+                (short) index, (short) log2Group, (short) log2Delta,
+                (short) nDelta, isMultiPageSize, isSubpage, (short) log2DeltaLookup
+        };
+
+        sizeClasses[index] = sz;
+        int size = (1 << log2Group) + (nDelta << log2Delta);
+
+        if (sz[PAGESIZE_IDX] == yes) {
+            nPSizes++;
+        }
+        if (sz[SUBPAGE_IDX] == yes) {
+            nSubpages++;
+            smallMaxSizeIdx = index;
+        }
+        if (sz[LOG2_DELTA_LOOKUP_IDX] != no) {
+            lookupMaxSize = size;
+        }
+        return size;
+    }
+
+    private void idx2SizeTab() {
```

Comment:
can me merge this into the constructor or mark it as static and call it from the constructor and just pass in `sizeIdx2sizeTab` and `pageIdx2sizeTab ` ? This way we could define these two final and so the JVM / JIT may be able to optimise access.